File: functions/vincentiles_eeg.m
function [vin, vineeg, vin_index, vin_data] = vincentiles_eeg(dat,eegdata,nb_bins)

% vincentiles_eeg.m
%
% Computes nb_bins vincentiles for the data vector 'dat' (from condtrials.RT) and sorts and
% averages the eegdata according to the bins. Needs output from
% collectconds_bar13! eegdata shape must (!) be timepoint x trials
% 
% Originally written by:  Trisha van Zandt
% adapted by Sven Hoffmann for EEG-vincentiles 15/07/2013
% adapted by Saskia Wilken to also give indices and vincentiile data
% 30.11.2022


% sort data and get length and dimensions
[dat, index] = sort(dat);
n = length(dat);
neeg = size(eegdata,1);
dims = size(dat);
dimseeg = size(eegdata);
if dims(2) == 1,
   dat = dat';
end

% sort eegdata
tmpeeg=zeros(dimseeg);
for ind = 1:length(index)
   tmpeeg(:,ind) = eegdata(:,index(ind));
end


% create temporary arrays with nb_bins replications of each value
copyx = ones(nb_bins,1)*dat;
vinx = reshape(copyx,n,nb_bins);

copyxeeg = repmat(tmpeeg',[1,1,nb_bins]);

vinxeeg = shiftdim(copyxeeg,2);
vinxeeg = reshape(vinxeeg,dimseeg(2),nb_bins,neeg);

% obtain mean values for each (RT EEG)bin
vin = mean(vinx)';
vineeg = squeeze(mean(vinxeeg,1));

% get indices of trials per vincentile
index = index(1:floor(length(index)/nb_bins)*nb_bins); % omits some values
vin_index = reshape(index, [length(index)/nb_bins, nb_bins]);

% get data associated with vincentiles
vin_data = dat(vin_index);

File: functions/vincentiles_eeg.r
# vincentiles_eeg.r
#
# Computes nb_bins vincentiles for the rt vector 'dat' and corresponding eegdata. Sorts and
# averages the eegdata according to the bins. eegdata shape must (!) be timepoint x trials
# 
# Originally written by:  Trisha van Zandt
# input: 
# dat: rt (vector)
# eegdata: EEG data array (frames,trials)
# nb_bins: number of bins to compute
# adapted by Sven Hoffmann for EEG-vincentiles and translated to GNU R 15/07/2013

vincentiles_eeg <- function(dat,eegdata,nb_bins){
  # sort data and get length and dimensions
  sorted <- sort(dat,index=TRUE)
  dat <-  sorted$x
  index <-  sorted$ix
  ntrialseeg <- ncol(eegdata)
  nframeseeg <- nrow(eegdata)
  # sort eegdata according to rts
  tmpeeg <- matrix(0,nframeseeg,ntrialseeg)
  for (ind in c(1: length(index))){
    tmpeeg[,ind] <- eegdata[,index[ind]]  
  }
  
  # create temporary arrays with nb_bins replications of each value (vincentiles)
  #rts
  copyx <- rep(dat,each = nb_bins)
  vinx<-array(0,c(nb_bins,ntrialseeg))
  counter <- 1
  for (ind in c(1:nb_bins)){
    vinx[ind,] <- copyx[c(counter:(counter +ntrialseeg-1))]
    counter <-  counter + ntrialseeg
  }
  
  #eeg (eeg vincentiles according to corresponding rts)
  copyxeeg <- tmpeeg[, rep(1:ncol(tmpeeg), each = nb_bins)]
  vinxeeg<-array(0,c(nframeseeg,ntrialseeg,nb_bins))
  counter <- 1
  for (ind in c(1:nb_bins)){
    vinxeeg[,,ind] <- copyxeeg[,c(counter:(counter +ntrialseeg-1))]
    counter <-  counter + ntrialseeg
  }
  
  # obtain mean values for each (RT EEG) bin
  
  return(list(vince_eeg=apply(vinxeeg, c(1,3), mean), vince_rts=rowMeans(vinx)))
  
}